flocker/node/_diagnostics.py

```diff
@@ -0,0 +1,313 @@
+            )
+
+            # Network configuration
+            check_call(
+                ['ip', 'addr'],
+                stdout=self._open_logfile('ip-addr')
+            )
+
+            # Hostname
+            self._open_logfile('hostname').write(gethostname() + '\n')
+
+            # Create a single archive file
+            archive_path = make_archive(
+                base_name=self._archive_name,
+                format='tar',
+                root_dir=os.path.dirname(self._archive_path),
+                base_dir=os.path.basename(self._archive_path),
+            )
+        finally:
+            # Attempt to remove the source directory.
+            rmtree(self._archive_path)
+        return archive_path
+
+
+class SystemdServiceManager(object):
+    """
+    List services managed by Systemd.
+    """
+    def all_services(self):
+        """
+        Iterate the name and status of all services known to SystemD.
+        """
+        output = check_output(['systemctl', 'list-unit-files', '--no-legend'])
+        for line in output.splitlines():
+            line = line.rstrip()
+            service_name, service_status = line.split(None, 1)
+            yield service_name, service_status
+
+    def flocker_services(self):
+        """
+        Iterate the name and status of the Flocker services known to SystemD.
+        """
+        service_pattern = r'^(?P<service_name>flocker-.+)\.service'
+        for service_name, service_status in self.all_services():
+            match = re.match(service_pattern, service_name)
+            if match:
+                service_name = match.group('service_name')
+                if service_status == 'enabled':
+                    yield service_name, service_status
+
+
+class UpstartServiceManager(object):
+    """
+    List services managed by Upstart.
+    """
+    def all_services(self):
+        """
+        Iterate the name and status of all services known to Upstart.
+        """
+        for line in check_output(['initctl', 'list']).splitlines():
+            service_name, service_status = line.split(None, 1)
+            yield service_name, service_status
+
+    def flocker_services(self):
+        """
+        Iterate the name and status of the Flocker services known to Upstart.
+        """
+        for service_name, service_status in self.all_services():
+            if service_name.startswith('flocker-'):
+                yield service_name, service_status
+
+
+class JournaldLogExporter(object):
+    """
+    Export logs managed by JournalD.
+    """
+    def export_flocker(self, service_name, target_path):
+        """
+        Export logs for ``service_name`` to ``target_path`` compressed using
+        ``gzip``.
+        """
+        # Centos-7 doesn't have separate startup logs.
+        open(target_path + '_startup.gz', 'w').close()
+        check_call(
+            'journalctl --all --output cat --unit {}.service '
+            '| gzip'.format(service_name),
```

Comment:
I was trying to reduce the amount of disk space used by the tool. 
We know that our log files are huge and nodes have been known to run out of disk space, so I tried not to make things worse.

Ideally the `tarfile` or `zipfile` modules would give you a file like object that could be supplied to `check_call(stdout=...` but there doesn't seem to be such an interface.

Nor could I find an option for the `tar` command to append the content of stdin as a named file in the archive.